### ISSUE TITLE
Extremely Repetitive Output

### ISSUE THREAD

**Issue by YaroKasear** [cloveranon/Clover-Edition#169]
I don't know if I set this up incorrectly or not, but it seems like no matter which model I use (I tried horni, horni-ln, and EleutherAI.) the output goes repetitive almost immediately, sometimes even only giving me output from things like my own custom context/prompt. 

It's output like, "You love your wife. You love your wife. You love your wife. You love your wife. You love your wife. You love your wife. You love your wife. You love your wife. You love your wife. You love your wife." 

Or even, "You turn off the TV. You stand up and leave the room. You turn on the TV. You turn off the TV. You love your wife. You turn off the TV." 

It weirdly just winds up deciding your input is all that matters. Sometimes it'll bother to actually use the model and output something new, but it'll just mix that in with all the repetition.

I checked the setting on repetitiveness and it's at the default. I don't understand, as I thought CE very explicitly was sold on this not being a thing in its models. I have AI Dungeon 2 Unleashed on the same machine with an RTX 3090 and its able to produce something usable. 

I haven't fine-tuned, but from what I understand, the "horni" models are already supposed to be plenty fine-tuned, aren't they? 

Please help, as I'm at the point where I might just get rid of Clover Edition and stick with AID2 Unleashed.

**Comment by MikkoMMM** [comment]
It's that bad? I've noticed some repetitiveness myself and thought it was a thing to work on later but I felt it was at least usable. @finetuneanon, you authored the GPT-Neo update so I'm pinging at least you.

You can find the old config.ini here:
https://raw.githubusercontent.com/cloveranon/Clover-Edition/6406d9a727bfd72a6e2fed210b01e7d1c107cdbb/config.ini

There were also two new configuration variables that you'll need to add. I think they can be disabled with:

```
#Repetitiveness Penalty Range
#controls how far back tokens are penalized
rep-pen-range = 0

#Repetitiveness Penalty Slope
#controls the slope of the penalty curve
rep-pen-slope = 3.33
```

Actually I don't think the latter one's value matters if the first is zero so I left it as is.

The old configuration was from a time before GPT-Neo, so I can mostly just suggest the "Original Clover Edition" model to use with it. There probably were reasons that the config was modified, after all.

**Comment by cloveranon** [comment]
If this isn't actually a bug, it could be caused by to low temp, top-k and top-p. All of which decrease the randomness of the AI, and it's default nonrandom behaviour is to just repeat for some reason (all text models do this, no one has a good explanation as to why.)

**Comment by MikkoMMM** [comment]
Yeah, the GPT-Neo support is pretty new still and we very well might not have the optimal configuration yet (not that there even is such a singular thing but you know). This may also be seen as an opportunity to experiment and help make the default configuration values better.

**Comment by finetuneanon** [comment]
> It's that bad? I've noticed some repetitiveness myself and thought it was a thing to work on later but I felt it was at least usable. @finetuneanon, you authored the GPT-Neo update so I'm pinging at least you.
> 
> There is some old configuration you could try, but it was from a time before GPT-Neo, so I can mostly just suggest the "Original Clover Edition" model to use with it. There probably were reasons that the config was modified, after all.

Using neo models with the old config did feel like it caused a lot of repetition. If increasing temperature and repetition penalty don't help much, I recommend increasing top_p, as it will often leave only a single choice at 0.9.

> Actually I don't think the latter one's value matters if the first is zero so I left it as is.

That's correct.

**Comment by YaroKasear** [comment]
I will work on tweaking the configs first to see if this helps before reverting to GPT-2 models. If I find myself with great success, I'll share my numbers.

**Comment by Skhmt** [comment]
increasing temp to 0.9 seems to help for me

**Comment by YaroKasear** [comment]
Temp adjustment helped. I also tweaked top-k and top-p. It's not yet close to being a "quality" experience, as I think I swung a bit far the other way and the game is "yes, anding" me.

**Comment by MikkoMMM** [comment]
Do we have updated values for our default configuration yet? It seems the config.ini hasn't been updated at least.

**Comment by Skhmt** [comment]
Its like not even science, just "feel" when testing. So far temp at 0.9 and penalty at 1.4 has made zero repeats, but that might have taken it too far? Haven't messed with top k and p at all.

**Comment by cloveranon** [comment]
I have some thoughts about it.

-  Ideally generator settings should probably be per model. This is not the way the code is set up at all though.
- Ideally we should have a way of "scientifically" testing different settings. E.g. generate a bunch of different outputs for different settings and prompts and manually A/B test them.
- I've been thinking that neo might not actually be that good.
